File: src/lib/scala/Slate.Api/src/main/scala/slate/core/apis/doc/Doc.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.core.apis.doc

import slate.common.console.ConsoleWriter
import slate.core.apis.{Api, ApiArg}

abstract class Doc extends ApiVisit {
  private val _writer = new ConsoleWriter()
  private val _settings = new DocSettings()
  import _writer._

  def settings:DocSettings =
  {
    _settings
  }


  def lineBreak():Unit =
  {
    text("---------------------------------------------------------------", endLine = true)
  }


  override def onVisitSeparator(): Unit =
  {
    line()
  }


  override def onAreasBegin():Unit =
  {
    lineBreak()
    title("supported areas: ", endLine = true)
    line()
  }


  override def onAreasEnd():Unit =
  {
    text("type '{area} ?' to list all apis in the area. ")
    url("e.g. sys ?" , endLine = true)
    lineBreak()
  }


  override def onAreaBegin(area:String):Unit =
  {
    highlight(area, endLine = true)
  }


  override def onAreaEnd(area:String):Unit =
  {
  }


  override def onApisBegin(area:String):Unit =
  {
    lineBreak()
    title("supported apis: ", endLine = true)
    line()
  }


  override def onApisEnd(area:String):Unit =
  {
    line()
    text("type {area}.{api} ? to list all actions on an api. ")
    url("e.g. sys.models ?" , endLine = true)
    lineBreak()
  }


  override def onApiEnd(api: Api): Unit =
  {
    line()
  }


  override def onArgEnd(arg: ApiArg): Unit =
  {
    line()
  }


  override def onApiActionSyntax(): Unit =
  {
    line()
    text("type {area}.{api}.{action} ? to list inputs for an action. ")
    url("e.g. sys.models.install ?" , endLine = true)
    lineBreak()
  }


  protected def getFormattedText(text:String, max:Int):String =
  {
    if (text.length == max)
      text
    else
      text + 0.until(max - text.length).foldLeft("")( (s, v) => s + " ")
  }
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/types/Doc.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.types

/**
 * Represents string content type/format information.
 *
 * Use cases:
 * 1. Provide type info on a string
 * 2. Provide intent that the string should be treated as file vs content ( See Content.kt )
 * 3. Provide a way for consumers to infer the intent of the string.
 *    e.g. the API server can determine that if a service returns a Doc instead of a string,
 *         then the Doc should be sent back as a File instead of a string
 */
class Doc(val name: String, val data:ByteArray, val raw: String?, val tpe: ContentType, val size: Long) {

    constructor(name:String, content: String, tpe: ContentType, size: Long) :
        this(name, content.toByteArray(), content, tpe, size)

    val text: String? by lazy { raw ?: String(data) }


    companion object {
        @JvmStatic
        val empty = Doc("", byteArrayOf(),"", ContentTypeText, 0)

        @JvmStatic
        fun text(name: String, content: String): Doc = Doc(name, content, ContentTypeText, content.length.toLong())

        @JvmStatic
        fun html(name: String, content: String): Doc = Doc(name, content, ContentTypeHtml, content.length.toLong())

        @JvmStatic
        fun json(name: String, content: String): Doc = Doc(name, content, ContentTypeJson, content.length.toLong())

        @JvmStatic
        fun csv(name: String, content: String): Doc = Doc(name, content, ContentTypeCsv, content.length.toLong())

        @JvmStatic
        fun prop(name: String, content: String): Doc = Doc(name, content, ContentTypeProp, content.length.toLong())

        @JvmStatic
        fun other(name: String, content: String, tpe: ContentType): Doc =
                Doc(name, content, tpe, content.length.toLong())
    }
}
